### PR TITLE
Use dill to pickle job functions (resolves #264)

### LINKED ISSUE #264: Allow jobs to wrap class functions, nested functions and lambda functions
┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-307)
┆Issue Number: TOIL-307

### PR DESCRIPTION
**Pull request by jpfeil**
continued from #505 
resolves #264

### CODE REVIEW

**Comment by jpfeil** [comment]
@benedictpaten

**Comment by hannes-ucsc** [comment]
I don't like how the function is always pickled but then only unpickled as a fall-back when the old mechanism fails. The pickle string may be huge and we need to defend against that. We need to thoroughly understand if we really need both mechanisms and how they interact with each other.

File: src/toil/job.py [hunk 1]
```diff
@@ -1457,17 +1459,19 @@ def __init__(self, userFunction, *args, **kwargs):
...
-        #TODO: Add dill support
-        #else use indirect method
+
         self.userFunctionModule = ModuleDescriptor.forModule(userFunction.__module__).globalize()
-        self.userFunctionName = str(userFunction.__name__)
-        self._args=args
-        self._kwargs=kwargs
+        self.pickledUserFunction = dill.dumps(userFunction)
+        pickledLen = len(self.pickledUserFunction)
+        assert len(self.pickledUserFunction) < 16000, \
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Don't use assert for this. Assertion are for catching errors you made.

File: src/toil/job.py [hunk 2]
```diff
@@ -1457,17 +1459,19 @@ def __init__(self, userFunction, *args, **kwargs):
...
+            'User function {} is too large to pickle: {}'.format(userFunction.__name__, pickledLen)
+        logger.info('{} pickled size: {}'.format(userFunction.__name__,
+                                                 pickledLen))
+        self._args = args
+        self._kwargs = kwargs
 
     def _getUserFunction(self):
-        userFunctionModule = self._loadUserModule(self.userFunctionModule)
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Now you've completely disabled hot deployment. This will not work.

**Comment by hannes-ucsc** [comment]
We are dealing with this the wrong way around. All of the tests test the special case of running on a single machine where hot deployment isn't necessary. This causes you to draw the wrong conclusions. Like today when you completely disabled hot-deployment and the tests still magically pass. However, in the real-world that change would break Toil for everyone using Mesos, GridEngine etc.

**Comment by hannes-ucsc** [comment]
@benedictpaten and I decided to post-pone this until after the 3.2 release.

**Comment by hannes-ucsc** [comment]
We now have integration tests that could help validate changes to—or affecting the—hot-deployment subsystem. I can give direction wrt what a good integration test for this would look like.

**Comment by DailyDreaming** [comment]
Closing this stale PR.  Especially since dill is currently implemented in Toil.  @benedictpaten @hannes-ucsc

**Pull request closed without merging**